src/app/api/booking/route.ts

```diff
@@ -9,107 +10,87 @@ import { getCurrentUser } from "@/lib/session";
       });
 
+    // find user
     const user = await db.query.users.findFirst({
       where: eq(users.id, currentUser!.id),
     });
 
-    if (user?.member === true) {
-      if (num_of_sessions > 2) {
-        return new Response("num_of_sessions cannot be greater than 2", {
-          status: 400,
-        });
-      } else {
-        for (const session of json) {
-          const gameSession = await db.query.gameSessions.findFirst({
-            where: eq(gameSessions.id, session.gamesessionId),
-          });
-
-          const time = new Date();
-          if (time > gameSession!.endTime || time < gameSession!.startTime) {
-            return new Response(null, { status: 400 });
-          }
-        }
-        if (user.remainingSessions < num_of_sessions) {
-          return new Response(null, { status: 400 });
-        }
-
-        await db.transaction(async (tx) => {
-          await tx
-            .update(users)
-            .set({
-              remainingSessions: user.remainingSessions - num_of_sessions,
-            })
-            .where(eq(users.id, currentUser!.id));
-          await tx.insert(bookings).values({
-            userId: currentUser!.id,
-            gameSessionId: json[0].gamesessionId,
-            createdAt: new Date(),
-            difficulty: json[0].playLevel,
-          });
-          await tx.insert(bookings).values({
-            userId: currentUser!.id,
-            gameSessionId: json[1].gamesessionId,
-            createdAt: new Date(),
-            difficulty: json[1].playLevel,
-          });
-        });
-
-        return new Response("success", { status: 204 });
-      }
-    } else {
-      if (num_of_sessions !== 1) {
-        return new Response("num_of_sessions is not equal to 1", {
-          status: 400,
-        });
-      } else {
+    if (numOfSessions <= (user?.member ? 2 : 1) && numOfSessions > 0) {
+      // iterate through all the sessions in the request
+      for (const session of json) {
         const gameSession = await db.query.gameSessions.findFirst({
-          where: eq(gameSessions.id, json[0].gamesessionId),
+          where: eq(gameSessions.id, session.gameSessionId),
         });
 
+        // check if the booking is within the time period
         const time = new Date();
         if (time > gameSession!.endTime || time < gameSession!.startTime) {
-          return new Response(null, { status: 400 });
+          return new Response("booking not within the time period", {
+            status: 400,
+          });
         }
-        if (user!.remainingSessions < num_of_sessions) {
-          return new Response(null, { status: 400 });
+
+        // check if the session is already full
+        const booking = await db.query.bookings.findMany({
+          where: eq(bookings.gameSessionId, session.gameSessionId),
+        });
+        if (booking.length >= gameSession!.maxUsers) {
+          return new Response("game session is full", { status: 400 });
         }
+      }
```

Comment:
while this works, its actually suspectible to race conditions. Imagine we have capacity of 40 and at 39 bookings currently. if person A and B both make a booking at the same time:

- Person A checks is full: 39 bookings, proceeds
- Person B checks is full: 39 bookings, proceeds
- Booking made for person A -- now at 40 bookings
- Booking made for person B -- now at 41 bookings

you might want to research how to prevent this

Reply:
i will go watch some f1 races and see how they be handling those race conditions! 🔥